core/drivers/clk/clk.c

```diff
@@ -0,0 +1,263 @@
+{
+	free(clk);
+}
+
+static bool clk_check(struct clk *clk)
+{
+	if (!clk->ops)
+		return false;
+
+	if (clk->ops->set_parent && !clk->ops->get_parent)
+		return false;
+
+	if (clk->num_parents > 1 && !clk->ops->get_parent)
+		return false;
+
+	return true;
+}
+
+static void clk_compute_rate(struct clk *clk)
+{
+	unsigned long parent_rate = 0;
+
+	clk->rate = 0;
+
+	if (clk->parent)
+		parent_rate = clk->parent->rate;
+
+	if (clk->ops->get_rate)
+		clk->rate = clk->ops->get_rate(clk, parent_rate);
+	else
+		clk->rate = parent_rate;
+}
+
+size_t clk_get_num_parents(struct clk *clk)
+{
+	return clk->num_parents;
+}
+
+struct clk *clk_get_parent_by_index(struct clk *clk, size_t pidx)
+{
+	if (pidx >= clk->num_parents)
+		return NULL;
+
+	return clk->parents[pidx];
+}
+
+static void clk_init_parent(struct clk *clk)
+{
+	size_t pidx = 0;
+
+	switch (clk->num_parents) {
+	case 0:
+		break;
+	case 1:
+		clk->parent = clk->parents[0];
+		break;
+	default:
+		pidx = clk->ops->get_parent(clk);
+		if (pidx >= clk->num_parents)
+			panic("get_parent returned an invalid value");
+
+		clk->parent = clk->parents[pidx];
+		break;
+	}
+}
+
+TEE_Result clk_register(struct clk *clk)
+{
+	if (!clk_check(clk))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	clk_init_parent(clk);
+	clk_compute_rate(clk);
+
+	DMSG("Registered clock %s, freq %ld", clk->name, clk_get_rate(clk));
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result clk_enable_no_lock(struct clk *clk)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct clk *parent;
+
+	if (refcount_inc(&clk->enabled_count))
+		return TEE_SUCCESS;
+
+	parent = clk_get_parent(clk);
+	if (parent)
+		clk_enable(parent);
+
+	if (clk->ops->enable) {
+		res = clk->ops->enable(clk);
+		if (res)
+			return res;
+	}
+
+	refcount_set(&clk->enabled_count, 1);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_enable(struct clk *clk)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	cpu_spin_lock(&clk->lock);
+	res = clk_enable_no_lock(clk);
+	cpu_spin_unlock(&clk->lock);
+
+	return res;
+}
+
+static void clk_disable_no_lock(struct clk *clk)
+{
+	struct clk *parent = NULL;
+
+	if (!refcount_dec(&clk->enabled_count))
+		return;
+
+	if (clk->ops->disable)
+		clk->ops->disable(clk);
+
+	parent = clk_get_parent(clk);
+	if (parent)
+		clk_disable(parent);
+}
+
+void clk_disable(struct clk *clk)
+{
+	cpu_spin_lock(&clk->lock);
+	clk_disable_no_lock(clk);
+	cpu_spin_unlock(&clk->lock);
+}
+
+static TEE_Result clk_set_rate_no_lock(struct clk *clk, unsigned long rate)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned long parent_rate = 0;
+
+	if (clk->parent)
+		parent_rate = clk->parent->rate;
+
+	res = clk->ops->set_rate(clk, rate, parent_rate);
+	if (res)
+		return res;
+
+	clk_compute_rate(clk);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_set_rate(struct clk *clk, unsigned long rate)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (!clk->ops->set_rate)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	cpu_spin_lock(&clk->lock);
+	if (rate != clk->rate) {
```

Comment:
If the resulting rate is rounded this test may always be `true`. Perhaps that caller already knows that and always uses a good multiple. I wonder if this test is needed at all. If makes sense to keep it, what about the `if (clk->flags & CLK_SET_RATE_GATE && clk_is_enabled(clk))`, shouldn't that always be tested? I mean it seems to suggest that `clk_set_rate()` was called in a bad state.

Reply:
Indeed, I removed the `rate != clk-rate` check. Maybe I will reintroduce that later when rounding rate operation will be supported to actually check that we really are going to set the same rate.

Reply:
OK